File: 2390-removing-stars-from-a-string/2390-removing-stars-from-a-string.js
/**
 * @param {string} s
 * @return {string}
 */
class Stack{
    #arr;

    constructor(){
        this.#arr = [];
    }

    push(element){
        this.#arr.push(element);
    }

    pop(){
        this.#arr.pop();
    }

    top(){
        return this.#arr[this.#arr.length -1];
    }

    isEmpty(){
        return this.#arr.length == 0;
    }
}
var removeStars = function(s) {
    let st = new Stack();
    let i = 0;
    while(i < s.length){
        if(s[i] != "*"){
            st.push(s[i]);
        }else if(s[i] == "*" && !st.isEmpty()){
            st.pop();
        }
        i++;
    }

    let result = "";
    while(!st.isEmpty()){
        let ch = st.top();
        st.pop();
        result = ch + result;
    }

    return result;
};

File: 2390-removing-stars-from-a-string/2390-removing-stars-from-a-string.java
class Solution {
    public String removeStars(String s) {
        Stack<Character> st = new Stack<>();
            int i = 0;
            while(i < s.length()) {
                if(s.charAt(i) != '*') {
                    st.push(s.charAt(i));
                } else if(s.charAt(i) == '*' && !st.isEmpty()) {
                    st.pop();
                }
                i++;
            }
            StringBuilder result = new StringBuilder("");  // here it string co catenation is not used because it is expensive opeartion in java
            while(!st.isEmpty()) {
                result.append(st.peek());
                st.pop();
            }
            return result.reverse().toString();
    }
}